presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroup.java

```diff
@@ -838,54 +814,78 @@ private void queryFinished(ManagedQueryExecution query)
                 while (group.parent.isPresent()) {
-                    group.parent.get().descendantQueuedQueries--;
+                    group.parent.get().descendantQueuedQueries.decrementAndGet();
                     group = group.parent.get();
                 }
             }
             updateEligibility();
         }
     }
 
-    // Memory usage stats are expensive to maintain, so this method must be called periodically to update them
-    protected void internalRefreshStats()
+    private void updateWaitingQueuedQueries()
     {
-        checkState(Thread.holdsLock(root), "Must hold lock to refresh stats");
-        synchronized (root) {
-            if (subGroups.isEmpty()) {
-                cachedMemoryUsageBytes = 0;
-                for (ManagedQueryExecution query : runningQueries) {
-                    cachedMemoryUsageBytes += query.getUserMemoryReservation().toBytes();
+        // For leaf group, when no queries can run, all queued queries are waiting for resources on this resource group.
+        if (subGroups.isEmpty()) {
+            waitingQueuedQueries.set(queuedQueries.size());
+        }
+        else {
+            // For internal groups, when no queries can run, only queries that could run on its subgroups are waiting for resources on this group.
+            int waitingQueuedQueryCount = 0;
+            for (InternalResourceGroup subGroup : subGroups.values()) {
+                if (subGroup.canRunMore()) {
+                    waitingQueuedQueryCount += min(subGroup.getQueuedQueries(), subGroup.getHardConcurrencyLimit() - subGroup.getRunningQueries());
                 }
-                Optional<ResourceGroupRuntimeInfo> resourceGroupRuntimeInfo = getAdditionalRuntimeInfo();
-                resourceGroupRuntimeInfo.ifPresent(groupRuntimeInfo -> cachedMemoryUsageBytes += groupRuntimeInfo.getMemoryUsageBytes());
             }
-            else {
-                for (Iterator<InternalResourceGroup> iterator = dirtySubGroups.iterator(); iterator.hasNext(); ) {
-                    InternalResourceGroup subGroup = iterator.next();
-                    long oldMemoryUsageBytes = subGroup.cachedMemoryUsageBytes;
-                    cachedMemoryUsageBytes -= oldMemoryUsageBytes;
-                    subGroup.internalRefreshStats();
-                    cachedMemoryUsageBytes += subGroup.cachedMemoryUsageBytes;
-                    if (!subGroup.isDirty()) {
-                        iterator.remove();
-                    }
-                    if (oldMemoryUsageBytes != subGroup.cachedMemoryUsageBytes || subGroup.isDirty.get()) {
-                        subGroup.updateEligibility();
-                        subGroup.isDirty.set(false);
-                    }
+            waitingQueuedQueries.set(waitingQueuedQueryCount);
+        }
+    }
+
+    // Memory usage stats are expensive to maintain, so this method must be called periodically to update them
+    private void updateMemoryUsageInfo()
```

Comment:
Can you keep the lock guard check ?         checkState(Thread.holdsLock(root), "Must hold lock to refresh stats"); ?